Adding diff of OCI image metada/config file

sei-secheverria opened the pull request:
Addresses issue 274 by adding functionality to compare OCI image metadata files (config.json). More specifically:

 - Adds a module to compare metadata JSON files at the key level (not as raw files). Result is new section in summary.json, metadata_diffs, with the differences. Added functionality to be able to load and parse these metadata/config files.
 - Adds two sections to the summary field of the summary.json file: meta_failure_summary, and total_failure_summary. The first is the summary of the metadata file differences, the second is a combination of those and the previously existing file differences. This previously existing summary of differences by diffoscope on the files, which was called "failure_summary", is renamed to "file_failure_summary" for clarity.
 - Created new FailureSummary structure to better handle the different summaries in a consistent way.
 - Minor internal refactoring to remove duplication between functions that compare OCI spec image folders, and OCI runtime bundle folders.
 - Extracted umoci calls from diff_command into its own umoci module.
 - Minor refactoring of skopeo module to add new functionality to load configs without duplication.
 - Added make command to clean environment.
 - Added script to run vessel in docker more easily.
 - Updated readme with examples calling vessel to better align with actual folders where output ends up from docker runs.
 - Added some tentative new flags to config file to include known differences in OCI image metadata/config keys.

More information about changes:
 - Summary section of summary json file now looks like this (cut before the checksum summary):

```
    "summary": {
        "total_failure_summary": {
            "unknown_failures": 25,
            "flagged_failures": 220,
            "trivial_failures": 218,
            "nontrivial_failures": 2,
            "total_failures": 245
        },
        "file_failure_summary": {
            "unknown_failures": 24,
            "flagged_failures": 218,
            "trivial_failures": 216,
            "nontrivial_failures": 2,
            "total_failures": 242
        },
        "meta_failure_summary": {
            "unknown_failures": 1,
            "flagged_failures": 2,
            "trivial_failures": 2,
            "nontrivial_failures": 0,
            "total_failures": 3
        },
       ...
```
 - Metadata flags have this simplified structure, and are inside a new `metaflags` field in `diff_config.yaml`. If needed, we could add a "description" field as well.
 
```
  - id: '<issue id>'
    key: '<path to key>'
    severity: '[Low/High/Medium]'
```

- Metadata diffs look like this in `summary.json`, in new `meta_diffs` field. `value1` and `value2` show the different falues between config files (with `null` being used if key is not present in one of them), and `matched_flag` simply copies the flag that matched the key, if any (with `null` being used if it matched no meta flag). Note that, unlike regular diffs, only one flag at most can match a metadata diff, since keys are unique and are the define how to match the flag. (Reference for OCI image config spec here: https://specs.opencontainers.org/image-spec/config/#properties)
```
        {
            "key": "created",
            "value1": "2024-06-20T17:40:35.180132994Z",
            "value2": "2024-07-22T21:44:13.781194625Z",
            "matched_flag": {
                "category_id": "TIMEC001",
                "key": "created",
                "severity": "Low"
            }
        },
```

- The following, very tentative, issue ids were suggested. They follow the standard naming scheme, but add a "C" at the end of the string part, for now just to differentiate them more easily, and restart the numbering due to this, but we can change this. Note that, unlike the other issues in the model, some of these flags may not have very clear causes, and focus more on the fact that we know what the key is about, rather than the exact cause for the difference.

```
Timestamp changes due to creation time of image:
  - id: 'TIMEC001'
    key: 'created'
    severity: 'Low'

  - id: 'TIMEC002'
    key: 'history'
    severity: 'Low'

Changes in author creating the image:
  - id: 'ENVC001'
    key: 'author'
    severity: 'Low'

Changes in environment vars being set:
  - id: 'ENVC002'
    key: 'config/Env'
    severity: 'High'

Changes in how/where the entrypoint for a container is:
  - id: 'ENVC003'
    key: 'config/Entrypoint'
    severity: 'High'

  - id: 'ENVC003'
    key: 'config/Cmd'
    severity: 'High'

  - id: 'ENVC003'
    key: 'config/WorkingDir'
    severity: 'High'

Differences in platform definition:
  - id: 'PLATC001'
    key: 'architecture'
    severity: 'High'

  - id: 'PLATC001'
    key: 'os'
    severity: 'High'
```

sei-secheverria committed:
Skopeo: made more generic, added function to load config from image; also minor fixes to Dockerfile and added str repr for ImageURI

sei-secheverria committed:
OCI: added small abstraction to make it easier to get config from OCI image; also added error details to skopeo command failure

sei-secheverria committed:
Created module to compare OCI configs, not yet connected to main diff command

sei-secheverria committed:
Doc markings

sei-secheverria committed:
Renaming

sei-secheverria committed:
Config diff: added case where whole config field is not on one of the sides

sei-secheverria committed:
Config diff: fixed key capitalization, made comments clearer

sei-secheverria committed:
Fixed reame example, added vesssel diff script for easier use in docker, and added example of this to readme.

sei-secheverria committed:
Improving specificity of typehints for dicts, and fixed outdated comments

sei-secheverria committed:
Merge branch 'metadata' into feature/docker-metadata

# Conflicts:
#	vessel/utils/diffoscope.py

sei-secheverria committed:
Diffoscope: fixing comment and type hints

sei-secheverria committed:
Diff Configs: small update to receive Paths instead of strings; also adds to_dict method to ConfigDiff

sei-secheverria committed:
Diff Command: added call to get config diffs, created common method for writing summary file.

sei-secheverria committed:
OCI: minor variable clarifications

sei-secheverria committed:
Vessel container run script: added command to always build container first

sei-secheverria committed:
Separated umoci commands and handling to specific module

sei-secheverria committed:
Config diffs: variable name clarifications, more comments

sei-secheverria committed:
Diff command: clarification of some vars, restructuring to avoid duplication between compares, fixed bug when calling config compare from one of the two modes

sei-secheverria committed:
ConfigDiff: makes it JSON serializale

sei-secheverria committed:
Fixed bug where diff config was not writing the proper values to dict

sei-secheverria committed:
Merge branch 'metadata' into feature/docker-metadata

# Conflicts:
#	vessel/diff/diff_command.py
#	vessel/utils/diffoscope.py

sei-secheverria committed:
Moved modules around to better group diff modules, and prepare for adding new commands later

sei-secheverria committed:
Attempt to improve metadata diffs, not finished

sei-secheverria committed:
Venv: improved command to update lock file, added clean command

sei-secheverria committed:
Renaming

sei-secheverria committed:
MetadataDiff: restructured to compare whole config, any that are present, instead of only selected keys.

sei-secheverria committed:
MetadataDiff: fixing typo, adding separate function to compare dicts, and added unit tests

sei-secheverria committed:
Moved most files back to original folder to simplify integrating new changes from master (will redo later)

sei-secheverria committed:
Metadata diff: added support for flags and summaries

sei-secheverria committed:
Undid changes to start merging from master

sei-secheverria committed:
Merge branch 'main' into feature/docker-metadata

# Conflicts:
#	vessel/diff/diff_command.py

sei-secheverria committed:
Linting

sei-secheverria committed:
DiffCommand: extracted umoci call to external util

sei-secheverria committed:
DiffCommand: extracted common code from compare_images and compare_files

sei-secheverria committed:
DiffCommand: renamed unpack_images to convert_to_oci for clarity

sei-secheverria committed:
MetadataDiff: fixed bug, comment, added unit test

sei-secheverria committed:
MetadataDiff: added methods and integration to load flags, and to save results to summary json

sei-secheverria committed:
Metadata Diff: added sample config, minor bugfixes

sei-secheverria committed:
Quote consistency

sei-secheverria committed:
Diff: refatored to use common FailureSummary struct for all summaries, and also added separate total, file and metadata failure summaries

sei-secheverria committed:
Updated suggested issue categories to be more in line with current ones, added new for platform diffs

sei-aderr commented on README.md:
```diff
@@ -64,14 +64,17 @@ To see commands and options:
 
 Example running on two OCI tars:
-* `docker run --rm -v INPUTDIR:/input -v OUTPUTDIR:/output vessel diff oci-archive:/input/image1.tar oci-archive:/input/image2.tar -o /output`
+* `docker run --rm -v <INPUTDIR>:/opt/project/input -v <OUTPUTDIR>:/opt/project/output vessel diff -o ./output oci-archive:/input/image1.tar oci-archive:/input/image2.tar -o /output`
```
Is it intentional that there are now two `-o ./output`?

sei-secheverria replied:
Nope, good catch, one should be removed

sei-aderr committed:
Re-order diff_command functions

sei-aderr commented:
I did some re-ordering of the functions in `diff_command.py`. As we have added more functionality there, they have generally just been tacked onto the end and the order has gotten a bit free-form. I adjusted it to flow better, let me know what you think or if you have a different way you would order it

sei-aderr commented:
What do you think about breaking out the differences for `history` into separate diffs for each entry in that list? Currently it looks like it just prints out the entire history if any piece is different, but it isn't super easy to see what was different with that display.

sei-aderr commented on vessel/diff/failure.py:
```diff
@@ -0,0 +1,69 @@
+        return asdict(self)
+
+    @staticmethod
+    def calculate_file_failure_summary(
```
Is there a reason you structured this calculation like this and didn't have it be an `__init__` because it just ends up returning a `FailureSummary`

sei-secheverria replied:
Artifact of refactoring. There was a slight advantage of creating an empty one in some cases and calcualting the rest in another, but it is simpler to make minor changes where it is used and turned this into a constructor

sei-aderr commented on vessel/utils/metadata_diff.py:
```diff
@@ -0,0 +1,287 @@
+
+@dataclass
+class MetadataFlag:
+    """Represents a type of issue, and the ket it is being associated to."""
```
typo, `key`

sei-aderr commented on vessel/utils/metadata_diff.py:
```diff
@@ -0,0 +1,287 @@
+    """Compare dicts for differences."""
+    diffs: list[MetadataDiff] = []
+
+    # Checked keys is needed to avoid comparing keys that are in both dicts twice.
```
Wouldn't this just be keys that are in both dicts? They wouldn't both be in both dicts twice?

sei-secheverria replied:
The way the _compare_dicts_ref works is a little confusing, maybe I should look into making it clearer. Unlike diffoscope, which just compares lines, here the code is comparing things by keys. So we need a list of reference keys. But we would prefer not to have a "master" list of keys, even though the OCI image spec defines them, as sometimes you may find non-standard keys that come from other Docker specs and stuff, so it should be more robust to leave that open. But since the logic here compares by key, you need a list of keys to go over, and thus we chose the keys in one dict as the list of keys. But this means that, if there is a key in the second dict, it won't be checked, as we are using the list of keys from the first one as a reference. But if we do the comparison twice, the second time using the second dict as a reference, we would count those keys that are only in the second dict as part of the diffs. And to avoid the duplication, we used that checked_keys to avoid checking, the second time, the keys we already checked because they were in the first dict. The end result is that we end up checking the diffs for the keys of the first dict and second dict, without checking any twice.

I do not like this logic much, it is hard to follow. An alternative instead of doing this would be to first go get all the keys from one dict, then the other, then combine them, and then just do the pass once, with the combined list of keys as a reference. This would probably be much clearer.

sei-aderr replied:
Got it, yes it makes sense. The logic is good, I just misinterpreted what the commend was saying that it does. I like the idea of the combined dict of all the keys that you an pass over once, I have done that for similar cases in the past.

sei-secheverria replied:
Updated

sei-secheverria commented:
> Re-order diff_command functions

Yep, good reordering, looks much more logical right now.

sei-secheverria committed:
Typos

sei-aderr commented on vessel/utils/metadata_diff.py:
```diff
@@ -0,0 +1,287 @@
+    return flags
+
+
+def match_flags(
```
What do you think about specifying in the function name that this is for metadata flags? I am likely going to be adding something similar during the `parse_diffoscope` refactor for those flag matches. I see that when you call this you import the whole module, which does then specify where its coming from if you prefer that approach we just have to be sure we do that everywhere

sei-secheverria replied:
> What do you think about specifying in the function name that this is for metadata flags? I am likely going to be adding something similar during the `parse_diffoscope` refactor for those flag matches. I see that when you call this you import the whole module, which does then specify where its coming from if you prefer that approach we just have to be sure we do that everywhere

That is a good point. I tend to prefer importing modules and using that as a preface for each function, not only because it allows you to avoid making function names longer, but because, when reading code, it is much easier to see where it is coming from. For example, in diff_command, there are several functions that are currenty imported from checksum, but when going over the code is confusing to see if they are functions in the same diff_command or in an external module. (Sure, when clicking on it you'll know, but a quick parse is confusing). I have not changed it everywhere, but my preference would be to import modules and call the functions from there, unless it is a very generic util function or something like that.

sei-secheverria committed:
Dockerfile: updating pinned subversions of some packages

sei-aderr commented on vessel/utils/metadata_diff.py:
```diff
@@ -0,0 +1,287 @@
+                diff.matched_flag = flag
+
+    # Create summary of diffs.
+    summary = FailureSummary()
```
I both like this application of `FailureSummary()` to have a place to store the values instead of 3 separate variables for unknown, trivial, non-trivial, but find it odd because you aren't taking advantage of the one method of the class that is used to calculate the totals.

I am thinking it might make more sense to count unknown, trivial, non-trivial with this `FileSummary` and then return `calculate_file_failure_summary` of it but would like to hear your thoughts so we can decide on a best approach and I can model the refactors after it

sei-secheverria replied:
Yes, I just modified this as described above. Having the three variables as they are used in this method is not that terrible, but having the FailureSummary all over to encapsulate those three things is a benefit. As discussed above, just turned the calculate into init, and in this case, counted the three things separately and then passed them over to the FailureSummary init, which will calculate the other 2. This also helps having only one place/way to calculate the flagged/total bits.

sei-aderr replied:
Sorry to go back and forth with my comments, but do you think that we should maintain some way to store the 3 initial values in the class, and then have the ability for them to be calculated? It is nice to not have separate variables for each, but I don't have an immediate idea for how we could have an object for both the counting, and then the calculated values without separate classes. 

I would say I am quite in favor of having that separate class to simplify that group of 3 counts that is passed around a fair bit in the code if you don't mind implementing it as part of this PR.

sei-secheverria replied:
We could do both in the same class. We can have that calculate method be a separate method, that will be called automatically if the constructor is called, or can be called manually if you are setting each var separately. I'll push an update doing this and you can evaluate if it works for your purposes.

sei-aderr replied:
That does look like it will work, thank you!

Something I notice here, and I have had trouble with before in other pieces of the development of these tools is certain class structures ending up in somewhat of an ambiguous state as they are in the middle of being used/processed. This one is fairly easy, because you could always just recalculate the combined values and all will be well but it still doesn't feel super clear when those calculated values are there or not. The other example that comes to mind is some of the class variables for `DiffCommand` as they are used/populated throughout its execution but it never felt like I was able to make it clear when each had been populated and could be used, would still be at a default value that then needed to be set to manually, etc. Do you have any broad thoughts on how you handle situations like this?

sei-secheverria replied:
Not sure if there is one-size-fits-all solution for that issue. Some thoughts:
 - One option is to set clear default values, and be sure to check for those default values when needed. Sometimes this is enough, and sometimes it doesn't matter if the value is like so because it was the default or it was set to the invalid case for that value at some point, since the check is equivalent. You just have to have a clear "invalid" value. It does require you to make that check explicitly wherever it is needed, but it adds robustness to check for that case, since you should be doing it anyway. This would break down if you specifically want to check if the object is ready, as a separate check from invalid values.
 - Have one or more methods that make that check for you. Similar to above, but instead of having to remember the check, you just call a method that checks if it is valid/ready whatever. You can have one combined method if it makes sense, or several semi-combined or separate if it makes more sense. Still requires you to call the method to make the check.
 - Having one (or sometimes more) flag in the object that is set when the object is ready. You can then only explicitly make this check, and only go forward when this is done.  The issue here is figuring out how to set the flag, which in the end means either running some method to check and set it, or constantly do that check manually after each step. This makes sense if you have a combined state you want to check for, not for each attribute separately (you could have a flag per attribute, but that would be overkill).
 - What is commonly done in more object-oriented languages, and with minor workarounds can be done in Python, is to avoid accessing attributes directly, and do it through one or more methods instead. That can be a method per attribute, or a couple of methods that return several things. Thus, you can embed the check in that method, and you can forget about checking, you just have to use the correct accessor method.
 - Another potential option is similar to what you are saying, basically to compose an object based on others, and ensure the parent object is not ready if the children are not. This can be done in different ways: using None/null for the child objects and making that check; creating child objects separately, and only creating the parent one once you have all the parts; etc. In the first case, it is somewhat similar to the first couple of options, but it can be made clearer through the use of None/null; in the second case, you can't really create the full object until you have the pieces. These two options, and a bit more the second one, do add a lot of code, so it may make sense for more complex objects only.

sei-aderr replied:
Thank you for in the depth response. I had been thinking about going the getter/setter route with private class variables so will consider these options in the future.

sei-secheverria committed:
Dockerfile: removing uneeded dep that was causing trouble.

sei-secheverria committed:
FailureSummary: converted calculate method to init

sei-secheverria committed:
Linting

sei-secheverria commented:
> What do you think about breaking out the differences for `history` into separate diffs for each entry in that list? Currently it looks like it just prints out the entire history if any piece is different, but it isn't super easy to see what was different with that display.

Right. Thought about it, but not sure it is worth the effort, mostly due to how the diff is structured. We may want to add that, but it will add a certain level of complexity. These are the reasons:
 - The way the diff is structured, it only looks for diffs in dicts or subdicts. That is, finds a key and checks the differences in their values, or if the values are dicts as well, digs into that and compares inside again. The history value is different, it is a list of issues, so we would have to add support and some logic for comparing arrays/lists as well (assuming ordering?, etc). This can be done, but it is a substantial amount more of logic, which wouldn't be used much in the OCI image spec.
 - Along with the above, we only have a way to indicate "keys"; that is, a key or path, such as "created" or "config/Env". If we want to analyze entries in a list, we would have to define semantics to indicate that, such as "history[0]" for example? This, of course, can be done, it just would make things more complex: flag matching would be a bit more complex, as diff keys would not exactly match diff keys (ie, if the flag key is a not exactly correct "history/created", and the diff key is "history[0]/created". Again, could be done, just adds more complexity.
 -  Each entry in the history list has a "created" value which will always be different as it is a timestamp. From my initial logic, I was kinda ignring this, though currently it is not but having a Low severity flag for other timestamps. So this may no longer be a good reason to ignore this.
 
With all that being said, if we feel this adds value, I can add it.

sei-aderr commented:
> > What do you think about breaking out the differences for `history` into separate diffs for each entry in that list? Currently it looks like it just prints out the entire history if any piece is different, but it isn't super easy to see what was different with that display.
> 
> Right. Thought about it, but not sure it is worth the effort, mostly due to how the diff is structured. We may want to add that, but it will add a certain level of complexity. These are the reasons:
> 
> * The way the diff is structured, it only looks for diffs in dicts or subdicts. That is, finds a key and checks the differences in their values, or if the values are dicts as well, digs into that and compares inside again. The history value is different, it is a list of issues, so we would have to add support and some logic for comparing arrays/lists as well (assuming ordering?, etc). This can be done, but it is a substantial amount more of logic, which wouldn't be used much in the OCI image spec.
> * Along with the above, we only have a way to indicate "keys"; that is, a key or path, such as "created" or "config/Env". If we want to analyze entries in a list, we would have to define semantics to indicate that, such as "history[0]" for example? This, of course, can be done, it just would make things more complex: flag matching would be a bit more complex, as diff keys would not exactly match diff keys (ie, if the flag key is a not exactly correct "history/created", and the diff key is "history[0]/created". Again, could be done, just adds more complexity.
> * Each entry in the history list has a "created" value which will always be different as it is a timestamp. From my initial logic, I was kinda ignring this, though currently it is not but having a Low severity flag for other timestamps. So this may no longer be a good reason to ignore this.
> 
> With all that being said, if we feel this adds value, I can add it.

I will refer to @sei-kpitstick for his thoughts. I think it definitely adds value, but with the amount of work it sounds like it would take it is probably not worth implementing at this time.

sei-secheverria committed:
Modifies FailureSummary to allow manual recalculation of aggregated values

sei-secheverria committed:
MetadataDiff: simplified so that dict comparison is done by first merging keys into one set, then diffing.

sei-aderr commented:
All looks great to me. Thanks for your work on this Sebastian!

sei-secheverria committed:
MetadataDiff: created MetadataDiffs class to contain list of diffs, and convert to dict only when needed; and added output file to write those diffs to separately (same info as in summary, but for consistency with diffoscope and checksum outputs).

Also improved use of Path to avoid string concats and explicit use of path separator "/"

sei-secheverria committed:
Internal restructuring for better order

sei-secheverria committed:
Metadata Diffs: fixed constructor issues

sei-secheverria committed:
Diff command: minor refactoring to put config file name as constant

sei-secheverria committed:
DiffCommand: internal refactoring to reduce duplication between JSON and image folder modes

sei-secheverria committed:
Diff Command: moved metadata comparison out of common method, so JSON comparison can just receive loaded diffs

sei-secheverria committed:
MetadataDiff: added method to load diffs from JSON file. Missing unit test, plus code to actually get that input path

sei-secheverria committed:
DiffCommand: added code to load meta_diff JSON file as one of the files when doing a JSON level comparison. Not tested yet.

sei-secheverria committed:
DiffCommand: checked args checks for 3 files

sei-aderr commented on vessel/diff/diff_command.py:
```diff
@@ -96,25 +100,25 @@ def execute(self: "DiffCommand") -> bool:
             logger.error(
                 "At least 2 inputs required. Acceptable values are 2 image paths, "
-                "or 2 JSON files (diffoscope output and checksum metadata)"
+                "or 3 JSON files (diffoscope output and checksum metadata)"
```
Add `meta_diffs` file to this error message and the next one

sei-aderr replied:
I see you added `RuntimeError` with some error handling for some of your new cases. Could you adjust these two `logger.error` to use that format instead so they are consistent?

sei-secheverria replied:
Restructured flow of execute method to unify this way of handling things

sei-aderr commented on vessel/diff/diff_command.py:
```diff
@@ -198,62 +203,70 @@ def _setup(self: "DiffCommand") -> bool:
+            elif input_file_name == self.METADATA_DIFF_OUTPUT_FILENAME:
+                metadata_json_path = input_file
+            else:
+                diffoscope_json_path = input_file
```
I feel like if we are enforcing strict naming for two of the inputs that are made from the tool, we may as well do all 3.

sei-secheverria replied:
Done

sei-aderr commented on vessel/diff/diff_command.py:
```diff
@@ -198,62 +203,70 @@ def _setup(self: "DiffCommand") -> bool:
+            or metadata_json_path == ""
+        ):
+            raise RuntimeError(
+                "When providing JSON files, three files are needed, and one must be a checksum_metadata.json file, and another a meta_diffs.json file."
```
Going off previous comment, I think we should specify all 3 types in this error message

sei-secheverria replied:
Done

sei-aderr commented:
Just had a couple small comments, works great and I am a big fan of the refactors

sei-secheverria committed:
DiffCommand: made error messages for files more consistent, added requirement for diffoscope output to have specific name.

sei-secheverria committed:
DiffCommand: internal refactoring to unify error handling, and to make execute flow depending on mode clearer.

sei-aderr commented on vessel/diff/diff_command.py:
```diff
@@ -94,65 +93,59 @@ def execute(self: "DiffCommand") -> bool:
+            return False
 
-    def _setup(self: "DiffCommand") -> bool:
+    def _setup(self: "DiffCommand"):
```
If this is no longer going to return a bool, the execution path inside that can return false should be adjusted to raise a `RuntimeError`

sei-secheverria replied:
Fixed

sei-aderr replied:
Looks great, thanks

sei-secheverria committed:
Updated return path that was no longer valid

sei-aderr commented:
Got some conflicts with Lihan's branch being merged

sei-secheverria committed:
Merge branch 'main' into feature/docker-metadata

# Conflicts:
#	Dockerfile
#	vessel/diff/diff_command.py

sei-secheverria commented:
> Got some conflicts with Lihan's branch being merged

Fixed

sei-kpitstick commented:
@sei-secheverria I see "matched_flags" listed in meta_diffs.json -> do these get updated if run in JSON mode based on the current diff_config.json?

sei-secheverria commented:
> @sei-secheverria I see "matched_flags" listed in meta_diffs.json -> do these get updated if run in JSON mode based on the current diff_config.json?

Yes, they are ignored when loading meta_diffs.json in JSON mode, and they they are recalculated, similar to how the diffoscope output flags are re-evaluated.

sei-kpitstick reviewed (approved):
Tested, looks good
